### ISSUE TITLE
Not sure pre-compiled views are running or not

### ISSUE THREAD

**Issue by aggaton** [RazorGenerator/RazorGenerator#94]
I installed RazorGenerator.Mvc (2.4.2) and RazorGenerator.MsBuild (2.4.4) from nuGet, build the project. Checking in obj/CodeGen, I see all the generated files as expected. I then put a break-point in the razor source view and much to my surprise the code breaks there, indicating it runs the regular razor script rather than the pre-compiled. I verified the PrecompiledMvcEngine ViewEngine is first, the mapping list is 0 though. If I put a breakpoint in the generated view in obj/CodeGen, it indicates that code is not in the binary. If I delete the source view, the application fails, another indication the pre-compilation is not being picked up, this should be able to run without any razor views. Any idea where the chain may have jumped off? 

This is an example of the generated code, not sure if the namespace ASP is a problem or not.

Is there a way to check what pre-compiled views exists at run-time?

```
#pragma warning disable 1591
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

namespace ASP
{
    using System;
    using System.Collections.Generic;
    
    #line 1 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
    using System.Data;
    
    #line default
    #line hidden
    using System.IO;
    using System.Linq;
    using System.Net;
    using System.Text;
    using System.Web;
    using System.Web.Helpers;
    using System.Web.Mvc;
    using System.Web.Mvc.Ajax;
    using System.Web.Mvc.Html;
    using System.Web.Routing;
    using System.Web.Security;
    using System.Web.UI;
    using System.Web.WebPages;
    
    #line 2 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
    using TiPSViewCore.Models.Organization.PSIM;
    
    #line default
    #line hidden
    
    #line 3 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
    using TiPSViewCore.ViewModels.AlarmKB;
    
    #line default
    #line hidden
    
    [System.CodeDom.Compiler.GeneratedCodeAttribute("RazorGenerator", "2.0.0.0")]
    [System.Web.WebPages.PageVirtualPathAttribute("~/Areas/DB/Views/GenericTable/DataCurrent.cshtml")]
    public partial class _Areas_DB_Views_GenericTable_DataCurrent_cshtml : System.Web.Mvc.WebViewPage<GenericTableDataCurrent>
    {
        public _Areas_DB_Views_GenericTable_DataCurrent_cshtml()
        {
        }
        public override void Execute()
        {
            
            #line 5 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
  
    Response.ContentType = "text/xml";
    Layout = null;

            
            #line default
            #line hidden
WriteLiteral("\r\n<rows");

WriteAttribute("total_count", Tuple.Create(" total_count=\"", 214), Tuple.Create("\"", 245)
            
            #line 9 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
, Tuple.Create(Tuple.Create("", 228), Tuple.Create<System.Object, System.Int32>(Model.TotalCount
            
            #line default
            #line hidden
, 228), false)
);

WriteAttribute("pos", Tuple.Create(" pos=\"", 246), Tuple.Create("\"", 267)
            
            #line 9 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
, Tuple.Create(Tuple.Create("", 252), Tuple.Create<System.Object, System.Int32>(Model.Position
            
            #line default
            #line hidden
, 252), false)
);

WriteLiteral(">\r\n    <userdata");

WriteLiteral(" name=\"!ViewChanged\"");

WriteLiteral(">");

            
            #line 10 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                              Write((Model.ViewChanged.ToString()));

            
            #line default
            #line hidden
WriteLiteral("</userdata>\r\n    <userdata");

WriteLiteral(" name=\"!ViewSortColumn\"");

WriteLiteral(">");

            
            #line 11 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                                 Write((!string.IsNullOrEmpty(Model.ViewSortColumn) ? Model.ViewSortColumn : ""));

            
            #line default
            #line hidden
WriteLiteral("</userdata>\r\n    <userdata");

WriteLiteral(" name=\"!ViewSortDirection\"");

WriteLiteral(">");

            
            #line 12 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                                    Write((!string.IsNullOrEmpty(Model.ViewSortDirection) ? Model.ViewSortDirection : ""));

            
            #line default
            #line hidden
WriteLiteral("</userdata>\r\n    <userdata");

WriteLiteral(" name=\"!ViewError\"");

WriteLiteral(">");

            
            #line 13 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                            Write((!string.IsNullOrEmpty(Model.ViewError) ? Model.ViewError : ""));

            
            #line default
            #line hidden
WriteLiteral("</userdata>\r\n    <userdata");

WriteLiteral(" name=\"!CriteriaError\"");

WriteLiteral(">");

            
            #line 14 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                                Write((!string.IsNullOrEmpty(Model.CriteriaError) ? Model.CriteriaError : ""));

            
            #line default
            #line hidden
WriteLiteral("</userdata> \r\n    <userdata");

WriteLiteral(" name=\"!FilterError\"");

WriteLiteral(">");

            
            #line 15 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                              Write((!string.IsNullOrEmpty(Model.FilterError) ? Model.FilterError : ""));

            
            #line default
            #line hidden
WriteLiteral("</userdata>      \r\n");

            
            #line 16 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
    
            
            #line default
            #line hidden
            
            #line 16 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
     foreach (DataRow drRow in Model.GenericTableTable.Rows)
    { 

            
            #line default
            #line hidden
WriteLiteral("    <row");

WriteAttribute("id", Tuple.Create(" id=\"", 1048), Tuple.Create("\"", 1090)
            
            #line 18 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
, Tuple.Create(Tuple.Create("", 1053), Tuple.Create<System.Object, System.Int32>(drRow[Model.Table +"^ID"].ToString()
            
            #line default
            #line hidden
, 1053), false)
);

WriteLiteral(" RowChangeState=\"0\"");

WriteLiteral(">\r\n");

            
            #line 19 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
        
            
            #line default
            #line hidden
            
            #line 19 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
         foreach (PSIMColumn oPSIMColumn in Model.GenericTableDataColumns.AllColumns)
        {
            switch (oPSIMColumn.PSIMTable.Name.ToUpper())
            {
                case "MOC":

            
            #line default
            #line hidden
WriteLiteral("                    <cell />\r\n");

            
            #line 25 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                    break;
                default:

            
            #line default
            #line hidden
WriteLiteral("                    <cell ");

            
            #line 27 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                     Write(MvcHtmlString.Create(Model.GetCellAttributes(oPSIMColumn, drRow)));

            
            #line default
            #line hidden
WriteLiteral("><![CDATA[");

            
            #line 27 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                                                                                                 Write(Html.Raw(Model.CellValue(oPSIMColumn, drRow)));

            
            #line default
            #line hidden
WriteLiteral("]]></cell>\r\n");

            
            #line 28 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
                      break;
            }
        }

            
            #line default
            #line hidden
WriteLiteral("    </row>\r\n");

            
            #line 32 "..\..\Areas\DB\Views\GenericTable\DataCurrent.cshtml"
    }  

            
            #line default
            #line hidden
WriteLiteral("</rows>\r\n");

        }
    }
}
#pragma warning restore 1591
```

**Comment by davidebbo** [comment]
I don't think the breakpoint test verifies much. To properly test, try writing out the assembly the code is in, as in [here](https://github.com/davidebbo-test/RazorGeneratorMvc51/blob/3db380af0382d3e536ab2419ac405204eb170693/RazorGeneratorMvc5/Views/Home/Index.cshtml#L16).

**Comment by aggaton** [comment]
Alright I added an attribute to get the assembly as suggested to one of my views and this was returned

    <userdata name="!Assembly">@GetType().Assembly)</userdata>

    <userdata name="!Assembly">App_Web_ilbnbnpq, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null)</userdata>

Should this be the name of the mvc webapp dll? Is this an indication that the pre-compiled views are not run?

I also added razorgenerator.directives to each view directory with GeneratePrettyNames: true in it to have it generate correct namespace, and the files are definitely part of the compilation due to the numerous problems I had to fix. Apparently full qualified names have to be used on some of the Models.

**Comment by davidebbo** [comment]
Correct, that means it's not using precompiled. Not sure why, but you can try comparing with my test project: https://github.com/davidebbo-test/RazorGeneratorMvc51/, which does work.

**Comment by pranavkm** [comment]
Are the views compiled into the assembly? You can use reflector \ ILSpy to do this.

**Comment by aggaton** [comment]
Alright, so I loaded the web app assembly into reflector, and according to it the namespaces with the precompiled views are indeed all there. Example:

namespace TiPSViewWeb.Areas.DB.Views.GenericTable
{
    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/AddGenericTableRecord.cshtml")]
    public class AddGenericTableRecord : WebViewPage<AddGenericTableRecordResult>
    {
        // Methods
        public AddGenericTableRecord();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/ApproveAllChanges.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class ApproveAllChanges : WebViewPage<MOCResultManager>
    {
        // Methods
        public ApproveAllChanges();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/ApproveCellChanges.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class ApproveCellChanges : WebViewPage<MOCResultManager>
    {
        // Methods
        public ApproveCellChanges();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/Columns.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class Columns : WebViewPage<GenericTableDataColumns>
    {
        // Methods
        public Columns();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/CreateLimitOverrideTypeColumn.cshtml")]
    public class CreateLimitOverrideTypeColumn : WebViewPage<ColumnsCreationResult>
    {
        // Methods
        public CreateLimitOverrideTypeColumn();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/DataCurrent.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class DataCurrent : WebViewPage<GenericTableDataCurrent>
    {
        // Methods
        public DataCurrent();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/DataPending.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class DataPending : WebViewPage<GenericTableDataPending>
    {
        // Methods
        public DataPending();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/DataProjected.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class DataProjected : WebViewPage<GenericTableDataProjected>
    {
        // Methods
        public DataProjected();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/DataUpdate.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class DataUpdate : WebViewPage<GenericTableDataUpdateResults>
    {
        // Methods
        public DataUpdate();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/EnforceAll.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class EnforceAll : WebViewPage<MOCResultManager>
    {
        // Methods
        public EnforceAll();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/GenericTableFields.cshtml")]
    public class GenericTableFields : WebViewPage<GenericTableSelectFieldsResponse>
    {
        // Methods
        public GenericTableFields();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/GenericTableSelectFields.cshtml")]
    public class GenericTableSelectFields : WebViewPage<GenericTableSelectFieldsResponse>
    {
        // Methods
        public GenericTableSelectFields();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/GetAllGenericTableDataForTag.cshtml")]
    public class GetAllGenericTableDataForTag : WebViewPage<GetAllGenericTableDataForTag>
    {
        // Methods
        public GetAllGenericTableDataForTag();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/GetError.cshtml"), Dynamic(new bool[] { false, true }), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class GetError : WebViewPage<object>
    {
        // Methods
        public GetError();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/GetRationalizeData.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class GetRationalizeData : WebViewPage<GetRationalizationDataForColumn>
    {
        // Methods
        public GetRationalizeData();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/GridHeaderMenu.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class GridHeaderMenu : WebViewPage<GenericTableDataColumns>
    {
        // Methods
        public GridHeaderMenu();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/LimitCalculate.cshtml")]
    public class LimitCalculate : WebViewPage<LimitCalculationResults>
    {
        // Methods
        public LimitCalculate();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/PendingGenericTableRecordsForTag.cshtml")]
    public class PendingGenericTableRecordsForTag : WebViewPage<PendingGenericTableRecordsForTag>
    {
        // Methods
        public PendingGenericTableRecordsForTag();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/PriorityMatrix.cshtml")]
    public class PriorityMatrix : WebViewPage<PriorityMatrixRequest>
    {
        // Methods
        public PriorityMatrix();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/PriorityMatrixData.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class PriorityMatrixData : WebViewPage<PriorityMatrixData>
    {
        // Methods
        public PriorityMatrixData();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/PriorityRationalize.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class PriorityRationalize : WebViewPage<PriorityRationalizeResult>
    {
        // Methods
        public PriorityRationalize();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/PriorityRationalizeBatch.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class PriorityRationalizeBatch : WebViewPage<PriorityRationalizeBatchResult>
    {
        // Methods
        public PriorityRationalizeBatch();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/PriorityRationalizeGrid.cshtml")]
    public class PriorityRationalizeGrid : WebViewPage<PriorityRationalizeBatchResult>
    {
        // Methods
        public PriorityRationalizeGrid();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/RejectAllChanges.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class RejectAllChanges : WebViewPage<MOCResultManager>
    {
        // Methods
        public RejectAllChanges();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/RejectCellChanges.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class RejectCellChanges : WebViewPage<MOCResultManager>
    {
        // Methods
        public RejectCellChanges();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/SelectTag.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class SelectTag : WebViewPage<AddGenericTableRecordResult>
    {
        // Methods
        public SelectTag();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/SelectText.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class SelectText : WebViewPage<SelectTextResult>
    {
        // Methods
        public SelectText();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/TagComboFilterData.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class TagComboFilterData : WebViewPage<ComboFilterResult>
    {
        // Methods
        public TagComboFilterData();
        public override void Execute();
    }

    [GeneratedCode("RazorGenerator", "2.0.0.0"), PageVirtualPath("~/Areas/DB/Views/GenericTable/TemplateComboFilterData.cshtml")]
    public class TemplateComboFilterData : WebViewPage<ComboFilterResult>
    {
        // Methods
        public TemplateComboFilterData();
        public override void Execute();
    }

    [PageVirtualPath("~/Areas/DB/Views/GenericTable/Toolbar.cshtml"), GeneratedCode("RazorGenerator", "2.0.0.0")]
    public class Toolbar : WebViewPage<ToolbarResult>
    {
        // Methods
        public Toolbar();
        public override void Execute();
    }
}

**Comment by davidebbo** [comment]
@aggaton please wrap your code with ``` before and after so it's properly formatted.

You may need to debug it, or try to compare to my working sample app.

**Comment by aggaton** [comment]
So I created a new library project and inserted your source directly and got rid of the installed RazorGenerator.Mvc reference, so I could debug and see what is going on. The main problem seem to be that the _mappings structure is empty. i.e. the Exists() function in CompositePrecompiledMvcEngine always return false, so the precompiled views will never get picked.

What piece of the code is responsible for building that structure? And more importantly why is it not called?

**Comment by aggaton** [comment]
So looking at the Assembly=>CustomAttributes of one of the pre-compiled views, I don't see PageVirtualPathAttribute listed, so somehow that information is not attached to each view class, hence it is not being picked up in the constructor of PrecompiledMvcEngine to build the _mappings structure... Why would that information be stripped or not set properly?

**Comment by pranavkm** [comment]
Is the attribute present in the corresponding generated file under the obj/CodeGen directory?

**Comment by aggaton** [comment]
Yes, look at the example code up top

```
    [System.CodeDom.Compiler.GeneratedCodeAttribute("RazorGenerator", "2.0.0.0")]
    [System.Web.WebPages.PageVirtualPathAttribute("~/Areas/DB/Views/GenericTable/DataCurrent.cshtml")]
```

**Comment by aggaton** [comment]
Alright, I think I fixed the problem. There was a version discrepancy between system.web.mvc and system.web.webpages. Originally I was running mvc 3 and webpages 1. I removed all mvc and webpages references, installed mvc 4 and webpages 2. That seemed to have corrected the problem. I now get all 335 view mappings. Now the web app assembly shows up correctly 

    <userdata name="!Assembly">WebApp, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null)</userdata>

However why does it still break in the razor script, is it executing both?
